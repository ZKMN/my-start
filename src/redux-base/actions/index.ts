export * from "./commonFlow";
export * from "./homePageFlow";
