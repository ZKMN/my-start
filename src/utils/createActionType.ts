export enum RequestStatuses {
  Request = "REQUEST",
  Success = "SUCCESS",
  Failure = "FAILURE",
}
export enum XHRMethod {
  Get = "GET",
  Post = "POST",
  Put = "PUT",
  Patch = "PATCH",
  Delete = "DELETE",
}

export const createActionType = (
  method: XHRMethod,
  actionName: string,
  addFailureType?: boolean,
) => {
  const actionTypes = {
    REQUEST: `${method}_${actionName}_${RequestStatuses.Request}`,
    SUCCESS: `${method}_${actionName}_${RequestStatuses.Success}`,
    ...(addFailureType && { FAILURE: `${method}_${actionName}_${RequestStatuses.Failure}` }),
  }

  return actionTypes;
};
