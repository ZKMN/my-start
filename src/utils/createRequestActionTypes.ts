export enum RequestActionStatuses {
  Request = "REQUEST",
  Success = "SUCCESS",
  Failure = "FAILURE",
}

export enum XHRMethod {
  Get = "GET",
  Post = "POST",
  Put = "PUT",
  Patch = "PATCH",
  Delete = "DELETE",
}
export interface ICreateRequestActionTypes {
  REQUEST: string;
  SUCCESS: string;
  FAILURE?: string;
}

export const createRequestActionType = (method: XHRMethod, actionName: string) => `${method}_${actionName}`;

export const createRequestActionTypes = (method: XHRMethod, actionName: string, addFailureType?: boolean) => {
  const actionTypes = {
    REQUEST: `${createRequestActionType(method, actionName)}_${RequestActionStatuses.Request}`,
    SUCCESS: `${createRequestActionType(method, actionName)}_${RequestActionStatuses.Success}`,
    ...(addFailureType && { FAILURE: `${createRequestActionType(method, actionName)}_${RequestActionStatuses.Failure}` }),
  }

  return actionTypes;
};
