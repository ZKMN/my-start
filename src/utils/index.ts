export * from "./addActionsToSagas";
export * from "./createRequestAction";
export * from "./createRequestActionTypes";
export * from "./addActionsToSagas";
export * from "./sagaUtils";
