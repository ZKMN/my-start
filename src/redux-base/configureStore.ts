import { createStore, applyMiddleware } from "redux";
import createSagaMiddleware from "redux-saga";
import { routerMiddleware } from "react-router-redux";
import { BrowserHistory } from "history";
import { composeWithDevTools } from "redux-devtools-extension/developmentOnly";

import { LOGOUT } from './actions';
import rootSaga from "./sagas/rootSaga";
import rootReducer from "./reducers/rootReducer";

export const configureStore = (history: BrowserHistory) => {
  const sagaMiddleware = createSagaMiddleware();
  const reduxRouterMiddleware = routerMiddleware(history);

  let createRootReducer = rootReducer(history);

  createRootReducer = (state, action) => {
    if (action.type === LOGOUT) {
      return rootReducer(history)(undefined, action);
    }

    return rootReducer(history)(state, action);
  };

  const store = createStore(
    createRootReducer,
    composeWithDevTools(
      applyMiddleware(
        reduxRouterMiddleware,
        sagaMiddleware,
      ),
    ),
  );

  sagaMiddleware.run(rootSaga);

  return store;
};
