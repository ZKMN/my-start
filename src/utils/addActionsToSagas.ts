import { XHRMethod, ICreateRequestActionTypes } from "./createRequestActionTypes";

export const getActions = [] as string[];
export const postActions = [] as string[];
export const putActions = [] as string[];
export const patchActions = [] as string[];
export const deleteActions = [] as string[];
export const commonActions = [] as string[];

export const addActionsToSagas = (actionTypes: ICreateRequestActionTypes[]) => {
  actionTypes.forEach((actionType: ICreateRequestActionTypes) => {
    if (actionType.REQUEST.includes(XHRMethod.Get)) return getActions.push(actionType.REQUEST);
    if (actionType.REQUEST.includes(XHRMethod.Post)) return postActions.push(actionType.REQUEST);
    if (actionType.REQUEST.includes(XHRMethod.Put)) return putActions.push(actionType.REQUEST);
    if (actionType.REQUEST.includes(XHRMethod.Patch)) return patchActions.push(actionType.REQUEST);
    if (actionType.REQUEST.includes(XHRMethod.Delete)) return deleteActions.push(actionType.REQUEST);
    return null;
  });
};
