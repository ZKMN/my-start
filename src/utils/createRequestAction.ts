import { ICreateRequestActionTypes } from "./createRequestActionTypes";
import { StringifiableRecord } from 'query-string';

interface TActionCallbackResult {
  type: string | undefined;
  payload?: Record<string, unknown>;
}

interface IRequestPayload {
  payload?: any; 
  routeParams?: any;
  [key: string]: unknown; 
}

type TActionCallback = (payload: any) => TActionCallbackResult;

export interface IRequestAction {
  type: string | undefined;
  payload: Record<string, unknown>;
  endpoint: string;
  responseType?: string;
  queryParams?: StringifiableRecord;
  routeParams?: Record<string, unknown>;
  successCallback: TActionCallback;
  failureCallback?: TActionCallback;
}

export const createRequestAction = (
  actionType: ICreateRequestActionTypes,
  requestEndpoint: string,
  responseType?: string,
) => (
  requestPayload?: IRequestPayload,
): IRequestAction => {

  const { payload, routeParams, ...rest } = Object(requestPayload);

  const failureCallback = (payload: any) => ({
    type: actionType.FAILURE,
    payload,
  });
  const successCallback = (payload: any) => ({
    type: actionType.SUCCESS,
    payload,
  });

  const action: IRequestAction = {
    type: actionType.REQUEST,
    queryParams: { ...rest },
    routeParams,
    payload,
    endpoint: requestEndpoint,
    responseType,
    successCallback,
    ...(actionType.FAILURE && { failureCallback }),
  };

  return action;
};
