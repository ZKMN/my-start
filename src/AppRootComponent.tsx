import { Route } from 'react-router';
import { Provider } from 'react-redux';
import { BrowserHistory } from "history";
import { ConnectedRouter } from 'connected-react-router';

import { configureStore } from './redux-base/configureStore';
import { MainRoutes } from './routes/MainRoutes';

export interface IAppRootComponent {
  history: BrowserHistory,
  store: ReturnType<typeof configureStore>
}

export const AppRootComponent = ({
  history,
  store,
}: IAppRootComponent) => (
  <Provider store={store}>
    <ConnectedRouter history={history}>
      <Route
        render={({ location, history, match }) => (
          <MainRoutes
            location={location}
            history={history}
            match={match}
          />
        )}
      />
    </ConnectedRouter>
  </Provider>
);
