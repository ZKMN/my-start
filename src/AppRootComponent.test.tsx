import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

import { configureStore } from './redux-base/configureStore';
import { AppRootComponent, IAppRootComponent } from './AppRootComponent';
import { routerTestProps } from 'utils';

const { history } = routerTestProps();

let wrapper: ShallowWrapper;

const props: IAppRootComponent = {
  history,
  store: {
    subscribe: jest.fn(),
    dispatch: jest.fn(),
    getState: jest.fn(),
  } as unknown as ReturnType<typeof configureStore>,
};

describe('AppRootComponent', () => {
  beforeEach(() => {
    wrapper = shallow(<AppRootComponent {...props} />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('matches snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
