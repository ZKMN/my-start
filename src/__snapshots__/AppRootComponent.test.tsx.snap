// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppRootComponent matches snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [MockFunction],
      "getState": [MockFunction],
      "subscribe": [MockFunction],
    }
  }
>
  <Connect(ConnectedRouterWithContext)
    history={
      Object {
        "action": "POP",
        "back": [Function],
        "block": [Function],
        "createHref": [Function],
        "forward": [Function],
        "go": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Route
      render={[Function]}
    />
  </Connect(ConnectedRouterWithContext)>
</Provider>
`;
